src/Covenant/Type.hs

```diff
@@ -706,3 +713,132 @@ substitute index toSub = \case
+--              but when I do that it blows up lists? I think I have to be misunderstanding something
+bindVars :: forall (ann :: Type) (a :: Type). Count "tyvar" -> ([Doc ann] -> PrettyM ann a) -> PrettyM ann a
+bindVars count' act
+  | count == 0 = crossBoundary (act [])
+  | otherwise = crossBoundary $ do
+      here <- asks (view currentScope)
+      withFreshVarNames count $ \newBoundVars ->
+        local (over boundIdents (Map.insert here newBoundVars)) (act newBoundVars)
+ where
+   -- Increment the current scope
+   crossBoundary :: PrettyM ann a -> PrettyM ann a
+   crossBoundary = local (over currentScope (+ 1))
+
+   count :: Int
+   count = review intCount count'
+
+-- Bad name, but anyway, looks up the Doc for a pretty
+lookupRigid :: forall (ann :: Type). Int -> Index "tyvar" -> PrettyM ann (Doc ann)
+lookupRigid (ScopeBoundary -> scopeOffset) argIndex = do
+  let argIndex' = review intIndex argIndex
+  here <- asks (view currentScope)
+  asks (preview (boundIdents % ix (here + scopeOffset) % ix argIndex')) >>= \case
+    Nothing -> -- TODO: actual error reporting
+               error $ "Internal error: The encountered a variable at arg index "
+                       <> show argIndex'
+                       <> " with true level "
+                       <> show scopeOffset
+                       <> " but could not locate the corresponding pretty form at scope level " <> show here
+    Just res' -> pure res'
+
+prettyRenamedWithContext :: forall (ann :: Type). Renamed -> PrettyM ann (Doc ann)
+prettyRenamedWithContext = \case
+    Rigid offset index -> lookupRigid offset index
+    Unifiable i -> lookupRigid 0 i -- ok maybe 'lookupRigid' isn't the best name
+    Wildcard w64 i -> pure $ "_" <> viaShow w64 <> "#" <> pretty (review intIndex i)
+
+prettyCompTWithContext :: forall (ann :: Type). CompT Renamed -> PrettyM ann (Doc ann)
+prettyCompTWithContext (CompT count funArgs)
+    | review intCount count == 0 = prettyFunTy (NonEmpty.toNonEmpty funArgs )
+    | otherwise =  bindVars count $ \newVars -> do
+        funTy <- prettyFunTy (NonEmpty.toNonEmpty funArgs) -- easier to pattern match
+        pure $ mkForall newVars funTy
+
+mkForall :: forall (ann :: Type). [Doc ann] -> Doc ann ->  Doc ann
+mkForall tvars funTyBody = case tvars of
+  [] -> funTyBody
+  vars -> "forall" <+> hsep vars <> "." <+> funTyBody
+
+prettyFunTy :: forall (ann :: Type). NonEmpty (ValT Renamed) -> PrettyM ann (Doc ann)
+prettyFunTy (arg :| rest) = case rest of
+  [] -> ("!" <>) <$>  prettyArg arg
+  (a:as) -> (\x y -> x <+> "->" <+> y) <$> prettyArg arg <*> prettyFunTy (a :| as)
+ where
+   prettyArg :: ValT Renamed -> PrettyM ann (Doc ann)
+   prettyArg vt | isSimpleValT vt = prettyValTWithContext vt
+                | otherwise = parens <$> prettyValTWithContext vt
+
+-- I.e. can we omit parens and get something unambiguous? This might be overly aggressive w/ parens but that's OK
```

Comment:
Thunks need parens in all cases. There is a significant difference between `Integer` and `(!Integer)` as types, as the former is effectively strict, while the latter is lazy.